### PR TITLE
List attribute groups CQRS query using Doctrine EntityRepository

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add CQRS query to list Attribute Groups This handler is original because it uses a Doctrine Entity Repository (since it was already there) This query is required for the component that. list attributes to generate combinations You can either list the whole list of the list by product
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | ~
| How to test?      | CI green
| Possible impacts? | ~

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23794)

### CODE REVIEW

File: src/Core/Domain/Product/AttributeGroup/Attribute/QueryResult/Attribute.php [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    /**
+     * @var string
+     */
+    private $color;
+
+    /**
+     * @var string
+     */
+    private $name;
```
Comment (jolelievre) [inline src/Core/Domain/Product/AttributeGroup/Attribute/QueryResult/Attribute.php]: I though it might be interesting to have multi lang values here

File: src/Adapter/Attribute/Repository/AttributeRepository.php [hunk 2]
```diff
@@ -82,4 +84,128 @@ public function assertAllAttributesExist(array $attributeIds): void
...
         }
     }
+
+    /**
+     * @param ProductId $productId
+     * @param LanguageId $langId
+     *
+     * @return array[]
+     */
+    public function getAttributesInfoByProductId(ProductId $productId, LanguageId $langId): array
```
Comment (jolelievre) [inline src/Adapter/Attribute/Repository/AttributeRepository.php]: Should me move this service into the Product domain? (it's new in 178)
Like in `namespace PrestaShop\PrestaShop\Adapter\Product\AttributeGroup\Attribute\Repository`

File: src/Core/Domain/Product/AttributeGroup/Query/GetProductAttributeGroups.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+{
+    /**
+     * @var ProductId
+     */
+    private $productId;
+
+    /**
+     * @var LanguageId
+     */
+    private $languageId;
```
Comment (jolelievre) [inline src/Core/Domain/Product/AttributeGroup/Query/GetProductAttributeGroups.php]: I feel like we shouldn't limit the value to a Language (from a command point of view) but return localized values
Even if in the interface we will likely use only the context Language (no need to switch these data when we switch language in the page since these values are not editable)

I had the same question for combination list (which has the same principle), but for combinations maybe it would be interesting for the merchant to see the Combination name change Although it might be harder for him to identify the combination he is searching for (but he can always switch back to his favorite language)

File: src/Core/Domain/Product/AttributeGroup/Attribute/QueryResult/Attribute.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+     */
+    public function getColor(): string
+    {
+        return $this->color;
+    }
+
+    /**
+     * @return string
+     */
+    public function getName()
```
Comment (Progi1984) [inline src/Core/Domain/Product/AttributeGroup/Attribute/QueryResult/Attribute.php]:
```suggestion
    public function getName(): string
```

**Review by Progi1984 — COMMENTED** [review]
Small comments

File: src/Adapter/Attribute/Repository/AttributeRepository.php [hunk 5]
```diff
@@ -82,4 +84,128 @@ public function assertAllAttributesExist(array $attributeIds): void
...
+    {
+        $associatedProductAttributes = $this->getProductAttributes($productId->getValue());
+        $attributeIds = array_unique(array_map(function (array $attributeByCombination): int {
```
Comment (PierreRambaud) [inline src/Adapter/Attribute/Repository/AttributeRepository.php]: It will maybe be faster to use a new methods `$this->getProductAttributesIds($productId->getValue());` for performance, I'm afraid of side effect if you have 1000 combinations, array_map is rarely fast

File: src/Adapter/Attribute/Repository/AttributeRepository.php [hunk 6]
```diff
@@ -82,4 +84,128 @@ public function assertAllAttributesExist(array $attributeIds): void
...
+     * @param int[] $combinationIds
+     * @param LanguageId $langId
+     *
+     * @return array<int, array<int, mixed>>
+     */
+    public function getAttributesInfoByCombinationIds(array $combinationIds, LanguageId $langId): array
+    {
+        $attributeCombinationAssociations = $this->getAttributeCombinationAssociations($combinationIds);
+
+        $attributeIds = array_unique(array_map(function (array $attributeByCombination): int {
```
Comment (PierreRambaud) [inline src/Adapter/Attribute/Repository/AttributeRepository.php:119]: Look similar to above

File: src/Adapter/Attribute/Repository/AttributeRepository.php (hunk 5 above)
Comment (PierreRambaud) [reply src/Adapter/Attribute/Repository/AttributeRepository.php]: I see that you use `getProductAttributes` only in this class, why not doing the uniq check in SQL, it will be faster

Comment (jolelievre) [reply src/Adapter/Attribute/Repository/AttributeRepository.php]: You mean a dedicated SQL query to retrieve the attribute IDs via leftJoin query?

File: src/Adapter/Attribute/Repository/AttributeRepository.php (hunk 6 above)
Comment (jolelievre) [reply src/Adapter/Attribute/Repository/AttributeRepository.php:119]: Yes in this case it is necessary we need both side of the association

For product however I simplified the query into a single SQL one

File: tests/Integration/Behaviour/Features/Scenario/Attribute/list_attribute_groups.feature [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    And attribute "orange" named "Orange" in en language exists
+    And attribute "blue" named "Blue" in en language exists
+    And attribute "green" named "Green" in en language exists
+    And attribute "yellow" named "Yellow" in en language exists
+    And attribute "brown" named "Brown" in en language exists
+    And attribute "pink" named "Pink" in en language exists
+
+  # NOTE: we don't initialize specific data since we already have the core features
+  Scenario: List attribute groups
+    When I list all attribute groups I should get following results:
```
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Scenario/Attribute/list_attribute_groups.feature]: the action formulation is wrong. (it is mixed between action and the outcome).
It should be like this
```
Given|When|Then I should see a following list of attributes
```

could be easily fixable in another PR though

File: tests/Integration/Behaviour/Features/Scenario/Attribute/list_attribute_groups.feature [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  Scenario: List attribute groups with attributes
+    When I list all attribute groups, the group "size" should have the following attributes:
+      | name[en-US] | color | position | reference |
+      | S           |       | 0        | s         |
+      | M           |       | 1        | m         |
+      | L           |       | 2        | l         |
+      | XL          |       | 3        | xl        |
+
+  Scenario: List attribute groups with attributes
+    When I list all attribute groups, the group "color" should have the following attributes:
```
Comment (zuk3975) [inline tests/Integration/Behaviour/Features/Scenario/Attribute/list_attribute_groups.feature]: Once above mentioned action is adjusted, we can just remove the first and second scenario and instead cover it here in context. Although this scenario won't have any outcome, as everything will be covered with "given" :man_shrugging: 
```
Given I see a following list of attributes:
...
And I see following list of attribute groups with attributes:
...
And the group "color" should have the following attributes:
...
```

**Review by Progi1984 — APPROVED** [review]

File: src/Adapter/Product/AttributeGroup/QueryHandler/AbstractAttributeGroupQueryHandler.php [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            foreach ($attributeGroupEntity->getAttributeGroupLangs() as $attributeGroupLang) {
+                $localizedNames[$attributeGroupLang->getLang()->getId()] = $attributeGroupLang->getName();
+                $localizedPublicNames[$attributeGroupLang->getLang()->getId()] = $attributeGroupLang->getPublicName();
+            }
+
+            $attributes = null;
+            if ($withAttributes) {
+                $attributes = [];
+                /** @var AttributeEntity $attributeEntity */
+                foreach ($attributeGroupEntity->getAttributes() as $attributeEntity) {
```
Comment (matks) [inline src/Adapter/Product/AttributeGroup/QueryHandler/AbstractAttributeGroupQueryHandler.php:76]: It's hard to read this function but it's not the PR author fault, it's because the data model is complex.

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
Thank you @zuk3975 and @jolelievre
